python/TestHarness/schedulers/RunPBS.py

```diff
@@ -0,0 +1,286 @@
+from QueueManager import QueueManager
+import os, re
+from TestHarness import util
+
+## This Class is responsible for maintaining an interface to the PBS scheduling syntax
+class RunPBS(QueueManager):
+    @staticmethod
+    def validParams():
+        params = QueueManager.validParams()
+        params.addParam('queue_template', os.path.join(os.path.abspath(os.path.dirname(__file__)), 'pbs_template'), "Location of the PBS template")
+        return params
+
+    def __init__(self, harness, params):
+        QueueManager.__init__(self, harness, params)
+        self.params = params
+
+    def getLaunchCommand(self, job_container):
+        """ Return appropriate PBS command depending on current QueueManager state """
+        test_unique = self.getUnique(job_container)
+
+        if self.checkStatusState():
+            job = self.getData(test_unique, job_id=True)
+            return 'qstat -xf %s' % (job['job_id'])
+        else:
+            job = self.getData(test_unique, template_data=True)
+            return 'qsub -h %s' % (job['template_data']['queue_script'])
+
+    def launchJob(self, job_container):
+        """ Perform necessary file copy and script creation. Execute job """
+        test_unique = self.getUnique(job_container)
+
+        # create and populate our template
+        template = self.populateTemplateParams(job_container, {})
+        self.putData(test_unique, template_data=template)
+
+        # Prepare the worker directory
+        self.copyFiles(job_container)
+
+        # Create the batch script
+        self.prepareQueueScript(template)
+
+        # Run command
+        output = util.runCommand(self.getLaunchCommand(job_container), cwd=template['working_dir'])
+        self.handleLaunch(job_container, output)
+
+    def checkJob(self, job_container):
+        """ Perform QSTAT and discover job status """
+        output = util.runCommand(self.getLaunchCommand(job_container))
+        self.handleQueueStatus(job_container, output)
+
+    def postLaunch(self, jobs):
+        """ Release jobs that were placed on hold """
+        print 'releasing jobs...'
+        # TODO, instead of one job at a time method, create a buffer of IDs to release at a time
+        for job_container in jobs:
+            test_unique = self.getUnique(job_container)
+            json_session = self.getData(test_unique, job_id=True)
+            util.runCommand('qrls %s' % (json_session['job_id']))
+
+    def populateTemplateParams(self, job_container, template):
```

Comment:
This method is OK here for now, but the expansion process is actually pretty neutral. I could be factored out into the QueueManager class and should be once we add a new queueing Scheduler.